diff -ur odr-0.6/lib/d_odr.f odr-0.6p/lib/d_odr.f
--- odr-0.6/lib/d_odr.f	2002-06-28 03:49:21.000000000 +0300
+++ odr-0.6p/lib/d_odr.f	2005-08-12 13:28:11.688928000 +0300
@@ -5343,7 +5343,8 @@
 C...EXTERNAL SUBROUTINES
       EXTERNAL
      +   DACCES,DCOPY,DEVJAC,DFLAGS,
-     +   DODLM,DODPCR,DODVCV,DUNPAC,DWGHT,DXMY,DXPY
+     +   DODLM,DODPCR,DODVCV,DUNPAC,DWGHT,DXMY,DXPY,
+     +   PRINTFCN
 
 C...INTRINSIC FUNCTIONS
       INTRINSIC
@@ -5907,7 +5908,10 @@
                END IF
                LUNR = LUNRPT
                DO 140 I=1,NPR
-                  CALL DODPCR(IPR,LUNR,
+                  IF (IPR2.GE.7) THEN
+                    CALL PRINTFCN(NITER, ACTRED, BETA, WSS)
+                  ELSE
+                    CALL DODPCR(IPR,LUNR,
      +                        HEAD,PRTPEN,FSTITR,DIDVCV,IFLAG,
      +                        N,M,NP,NQ,NPP,NNZW,
      +                        MSGB,MSGD, BETA,Y,LDY,X,LDX,DELTA,
@@ -5918,6 +5922,7 @@
      +                        WSS,RVAR,IDF,WORK(SD),
      +                        NITER,NFEV,NJEV,ACTRED,PRERED,
      +                        TAU,PNORM,ALPHA,F,RCOND,IRANK,INFO,ISTOP)
+                  END IF
                   IF (IPR2.GE.5) THEN
                      IPR = 2
                   ELSE
diff -ur odr-0.6/odrpack.py odr-0.6p/odrpack.py
--- odr-0.6/odrpack.py	2002-11-19 06:55:54.000000000 +0200
+++ odr-0.6p/odrpack.py	2005-08-12 13:28:11.708956800 +0300
@@ -988,7 +988,8 @@
                   [1, 1], # short, short
                   [2, 1], # long,  short
                   [1, 2], # short, long
-                  [2, 2]] # long,  long
+                  [2, 2], # long,  long
+                  [3, 0]] # function, none
 
         if self.rptfile is None and \
            (so_init is not None or
#diff -ur odr-0.6/setup.py odr-0.6p/setup.py
#--- odr-0.6/setup.py	2002-11-10 09:28:39.000000000 +0200
#+++ odr-0.6p/setup.py	2005-08-12 13:40:16.661387200 +0300
#@@ -13,7 +13,7 @@
# # separately-compiled LINPACK. Edit the Extension instances
# # below to reflect where your libraries are.
# 
#-opt_blas = 1
#+opt_blas = 0
# 
# # We need to compile FORTRAN files, so check if we would be
# # using gcc anyways: then we can use g77 with impunity.
#@@ -22,8 +22,8 @@
# cc = new_compiler()
# sysconfig.customize_compiler(cc)
# 
#-if (os.name == 'nt' and find_executable('gcc') is not None) or \
#-   (cc.compiler[0] == 'gcc'):
#+if (os.name == 'nt' and find_executable('gcc') is not None): 
#+# or (cc.compiler[0] == 'gcc'):
#     # we would be compiling with gcc; change it to g77
# 
#     old_cust_cc = sysconfig.customize_compiler
diff -ur odr-0.6/src/__odrpack.c odr-0.6p/src/__odrpack.c
--- odr-0.6/src/__odrpack.c	2002-11-10 05:09:04.000000000 +0200
+++ odr-0.6p/src/__odrpack.c	2005-08-12 13:28:11.718971200 +0300
@@ -1356,6 +1356,16 @@
   {NULL, NULL},
 };
 
+/* CALL PRINTFCN(NITER, ACTRED, BETA, WSS) */
+
+void printfcn_ (int *niter, double *actred, double *beta, double *wss)
+{
+    PyObject *arg01;
+    arg01 = Py_BuildValue("(idd)", *niter, *actred, *wss);
+
+    PyEval_CallObject (odr_global.fcn, arg01);
+}
+
 void init__odrpack()
 {
   PyObject *m, *d;
